pkg/kubelet/dockershim/helpers.go

```diff
@@ -393,3 +396,44 @@ type dockerOpt struct {
 func (d dockerOpt) GetKV() (string, string) {
 	return d.key, d.value
 }
+
+// sharedWriteLimiter limits the total output written across one or more streams.
+type sharedWriteLimiter struct {
+	delegate io.Writer
+	limit    *int64
+}
+
+func (w sharedWriteLimiter) Write(p []byte) (int, error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
+	limit := atomic.LoadInt64(w.limit)
```

Comment:
Note that it is possible to write past the limit. Imagine 2 write requests coming in simultaneously with only a few bytes left in the limit, both read the same limit, write their output, and then subtract from that limit. The end result is the limit has been passed.

If you care, you could use a `CompareAndSwapInt64` call to decrement the limit before writing to the delegate, but i'm not sure it's worth the trouble...

Reply:
Yes, you're right. I intended to keep this implementation simple, especially given that the LimitBytes API doesn't guarantee a hard limit (only a soft one). I don't think it's worth the trouble/complexity to implement a hard limit here.